clang/lib/Parse/ParseExpr.cpp

```diff
@@ -248,6 +248,25 @@ ExprResult Parser::ParseArrayBoundExpression() {
   // If we parse the bound of a VLA... we parse a non-constant
   // constant-expression!
   Actions.ExprEvalContexts.back().InConditionallyConstantEvaluateContext = true;
+  // For a VLA type inside an unevaluated operator like:
+  //
+  //   sizeof(typeof(*(int (*)[N])array))
+  //
+  // in which the expression N is supposed to be ODR-used, as is the `array`.
```

Comment:
```suggestion
  // N and array are supposed to be ODR-used.
```
@zyn0217 Do you have tests for that?

@AaronBallman for the VLA shenanigans

Reply:
It comes from a regression. Without https://github.com/llvm/llvm-project/pull/123883/commits/8dba2df946a367706605572f1f2935f70d6c16a6, 

```cpp
static constexpr int CONSTRAINED_METHOD_1 = 1;
static constexpr int CONSTRAINED_METHOD_2 = 2;

namespace constrained_member_sfinae {

template<int N> struct S {
  template<class T>
  static constexpr int constrained_method() requires (sizeof(int[N * 1073741824 + 4]) == 16) { // expected-warning {{variable length arrays in C++ are a Clang extension}} \
                                                                                                  expected-note {{value 4294967296 is outside the range of representable values of type 'int'}} \
                                                                                                  expected-note {{while calculating associated constraint of template 'constrained_method' here}}
    return CONSTRAINED_METHOD_1;
  }

  template<class T>
  static constexpr int constrained_method() requires (sizeof(int[N]) == 16);
};

template<>
template<typename T>
constexpr int S<4>::constrained_method() requires (sizeof(int[4]) == 16) {
  return CONSTRAINED_METHOD_2;
}

// Verify that there is no amiguity in this case.
static_assert(S<4>::constrained_method<double>() == CONSTRAINED_METHOD_2);

} // namespace constrained_member_sfinae
```

we will spew two warnings `variable length arrays in C++ are a Clang extension` on `sizeof(int[N * 1073741824 + 4]) == 16)` because it doubly transforms the UnaryExprOrTypeTraitExpr in RebuildUnaryExprOrTypeTrait.

Reply:
It's already there: https://github.com/llvm/llvm-project/blob/ebb27ccb08e0579825a53b218ff5b2ddc492626a/clang/test/SemaTemplate/concepts-out-of-line-def.cpp#L225-L248

Reply:
OH.
I see, thanks. Sorry, it's early morning here...